src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -307,44 +336,118 @@ private void EventLoop()
+                Interop.Error err = Interop.Sys.WaitForSocketEvents(_port, _buffer, &numEvents);
+                if (err != Interop.Error.SUCCESS)
+                {
+                    throw new InternalException(err);
+                }
 
-                    for (int i = 0; i < numEvents; i++)
+                var socketEvents = new ReadOnlySpan<Interop.Sys.SocketEvent>(_buffer, numEvents);
+                int batchedCount = 0;
+                foreach (var socketEvent in socketEvents)
+                {
+                    if (socketEvent.Data == ShutdownHandle)
+                    {
+                        shutdown = true;
+                    }
+                    else if (_handleToContextMap.TryGetValue(socketEvent.Data, out SocketAsyncContext context))
                     {
-                        IntPtr handle = _buffer[i].Data;
-                        if (handle == ShutdownHandle)
+                        if ((socketEvent.Events & Interop.Sys.SocketEvents.Error) != 0)
                         {
-                            shutdown = true;
+                            // there was an error, we use the non-buffered execution path (this should be very rare)
+                            context.HandleEvents(socketEvent.Events);
                         }
                         else
                         {
-                            Debug.Assert(handle.ToInt64() < MaxHandles.ToInt64(), $"Unexpected values: handle={handle}, MaxHandles={MaxHandles}");
-                            _handleToContextMap.TryGetValue(handle, out SocketAsyncContext context);
-                            if (context != null)
-                            {
-                                context.HandleEvents(_buffer[i].Events);
-                                context = null;
-                            }
+                            context.AddWaitingOperationsToBatch(socketEvent.Events, ioControlBlocks, batchedOperations, ref batchedCount);
                         }
                     }
                 }
 
-                FreeNativeResources();
-            }
-            catch (Exception e)
-            {
-                Environment.FailFast("Exception thrown from SocketAsyncEngine event loop: " + e.ToString(), e);
+                if (batchedCount > 0)
+                {
+                    int result = Interop.Sys.IoSubmit(_aioContext.Ring, batchedCount, _aioBlocksPointers);
+                    if (result != batchedCount)
+                    {
+                        Interop.Error lastError = Interop.Sys.GetLastError();
+                        throw new InternalException($"IoSubmit has failed with {lastError}, returned {result} instead {batchedCount}"); // todo: implement a while loop
+                    }
+
+                    // todo: perf: avoid the syscall by using a well known pattern that reads from the ring
+                    result = Interop.Sys.IoGetEvents(_aioContext.Ring, batchedCount, batchedCount, _aioEvents);
```

Comment:
important: it's possible to avoid this sys call like @tmds does that in https://github.com/tmds/Tmds.LinuxAsync/blob/11233fc8fc80862533e319f4d22855347981aace/src/Tmds.LinuxAsync/EPollAsyncEngine.LinuxAio.cs#L203

it was not hot on the profile, so it was sitting on my todo list